File: day_4/first_part.py
from typing import Iterable, List, Set


REQUIRED_FIELDS = {'byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid'}


class Passport(dict):
    
    @classmethod
    def parse_entry(cls, entry: List[str]) -> 'Passport':
        return cls({ 
            data.split(':')[0]: data.split(':')[1]
            for line in entry
            for data in line.split()
        })
            


class PassportVerifier:
    _required_fields: Set[str]

    def __init__(self, required_fields: Iterable[str]):
        self._required_fields = set(required_fields)

    def is_valid(self, passport: Passport) -> bool:
        missing_fields = set(self._required_fields)
        for field in passport.keys():
            missing_fields.discard(field)
        return len(missing_fields) == 0


def nb_passports_valid(passports: List[Passport], verifier: PassportVerifier) -> int:
    return sum(1 for passport in passports if verifier.is_valid(passport))


entries = []
line = input()
while line != '-':
    entry = []
    while line != '':
        entry.append(line)
        line = input()
    entries.append(Passport.parse_entry(entry))
    line = input()

passport_verifier = PassportVerifier(REQUIRED_FIELDS)

print("solution:", nb_passports_valid(entries, passport_verifier))

File: 2020/day_4/haskell/first_part.hs
module Main where

import System.IO
import qualified Data.Map as Map

type Passport = Map.Map String String
type PassportRule = ([Char], [Char] -> Bool)

main :: IO ()
main = do
  m <- doReadFile
  let passports = parsePassports m
  print $ length $ filter isValid passports
  return ()

requiredFields :: [PassportRule]
requiredFields = 
  [ ("byr", const True)
  , ("iyr", const True)
  , ("eyr", const True)
  , ("hgt", const True)
  , ("hcl", const True)
  , ("ecl", const True)
  , ("pid", const True)
  ]

isValid :: Passport -> Bool 
isValid p = isValid' p requiredFields

isValid' :: Passport -> [PassportRule] -> Bool 
isValid' p rf = 
  null rf ||
  case mvalue of
    Nothing -> False
    Just value -> check value && isValid' p (tail rf)
  where mvalue = Map.lookup field p
        field = (fst.head) rf
        check = (snd.head) rf

parsePassports :: [[Char]] -> [Passport]
parsePassports ps
  | null ps = [] 
  | otherwise = parsePassport ((words . unwords) passport) : parsePassports (tail rest)
      where (passport, rest) = break null ps

parsePassport :: [[Char]] -> Passport
parsePassport p 
  | null p = Map.empty
  | otherwise = Map.union (Map.fromList [(field, tail value)]) (parsePassport (tail p))
      where (field, value) = break (==':') (head p)

doReadFile :: IO [[Char]]
doReadFile = do
  withFile "input" ReadMode doReadLine

doReadLine :: Handle -> IO [[Char]]
doReadLine hFile = do
  isEnd <- hIsEOF hFile
  if isEnd
    then return []
    else do
      line <- hGetLine hFile
      rest <- doReadLine hFile
      return (line : rest)
